Update restock logic during refunds closes #29502

roykho opened the pull request:
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

This was worked on along with the help of @vedanshujain. We basically introduced a new meta to keep track of the count of stocks that are refunded and are opt in to restock. With this new count, we're able to get the difference in order to know how much stock to actually add or subtract from the line item whenever the "maybe_restock" method is called. We've decided to implement this fix for all new orders (WC 5.5+). Any orders before that will behave the same way as before.

Closes #29502 

### How to test the changes in this Pull Request:

PREREQ: run `composer dump-autoload`

1. Set https://github.com/woocommerce/woocommerce/blob/trunk/includes/class-woocommerce.php#L27 to version 5.5
2. Create an order with an item that keeps stocks.
3. Go to the order and do a refund on the line item reducing the stock by 1. Ensure to uncheck the box that says `Restock refunded items`.  Click the `Refund manually` button.
4. Note stock count for the product stays the same and not restocked.
5. On the same order click the `Update` button. Ensure stock count remains the same for the product and not restocked.
6. Manually edit the quantity of the product line item. Reduce it by 1. Click on `Save` button on the bottom. Ensure stock is increased for the product.
7. Click on `Update` button. Ensure stock count remains the same for the product and not restocked.
8. Manually edit the quantity of the product line item again and this time increase it by 1. Click on `Save` button on the bottom. Ensure stock is decreased for the product.
9. Click on `Update` button. Ensure stock count remains the same for the product and not restocked.

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Fix - Introduce meta to track stocks that refunded and restocked to properly handle stock recalculation.

roykho committed:
Update restock logic during refunds closes #29502

vedanshujain commented on includes/admin/wc-admin-functions.php:
```diff
@@ -206,17 +206,17 @@ function wc_maybe_adjust_line_item_product_stock( $item, $item_quantity = -1 ) {
+	$product                = $item->get_product();
+	$item_quantity          = wc_stock_amount( $item_quantity >= 0 ? $item_quantity : $item->get_quantity() );
+	$already_reduced_stock  = wc_stock_amount( $item->get_meta( '_reduced_stock', true ) );
+	$restock_refunded_items = wc_stock_amount( $item->get_meta( '_restock_refunded_items', true ) );
```
For existing orders with refunds, `_restock_refunded_items` won't be set, for those old orders, if someone clicks on the update, we will end up increasing the back any reduced stock on those old orders. 

Perhaps we can assume the value of  `$item_quantity - $already_reduced_stock` in case `_restock_refunded_items` meta is not set at all? (And set it to 0- when we update an order, if not already set).

roykho replied:
Unit tests already exists for this method `wc_maybe_adjust_line_item_product_stock` and it is passing.

vedanshujain replied:
Ah ok, but as you can see those tests were not enough to catch this bug. Should we still add some more?

roykho replied:
I am not sure if we can do anything for older orders. Because if `$restock_refunded_items` is not set, it is returning 0 already. So the logic already becomes `$item_quantity - $already_reduced_stock` as you can see here https://github.com/woocommerce/woocommerce/pull/29762/files#diff-0c9a45b990699b400112d411b46f124a4b420725af90363a568f85a7f44b98cbR219

roykho replied:
Additional unit tests added.

vedanshujain replied:
> Because if $restock_refunded_items is not set, it is returning 0 already. So the logic already becomes $item_quantity - $already_reduced_stock as you can see here

My bad, I meant initializing the $restock_refunded_itmes to `$refunded_item_quantity`. It will still be a bug, but a bit better I think.

Currently if you:
1. Refund an order with restocking in the older WooCommerce version (or trunk). See that stocks are increased for the product.
2. Click on update, nothing will happen.
3. Switch to a WooCommerce version with this fix, and click on update again. See that stock will be incorrectly reduced for the product. This is because `$restock_refunded_itmes` will return 0.

If we instead change `$restock_refunded_itmes` to be initialized to `refunded_item_quantity` in case its not present, it will be only as bad as the current (trunk) behavior, which is buggy for older orders only when stock is refunded without restocking.

roykho replied:
I am not sure if that would work correctly because while it would work for older orders that didn't have `$restock_refunded_items` but for new orders it is possible this is also not set yet. Then the same issue continues to happen for new orders.

roykho replied:
One thought I had is we handle both scenarios differently depending on the `order version`. If this order is from WC `<5.4`, we handle it the old way and if it is `>=5.4`, we handle it with the new logic?

vedanshujain replied:
>  If this order is from WC <5.4, we handle it the old way and if it is >=5.4, we handle it with the new logic

Ok, sounds good although it does look a bit hacky. I guess there is no better way? Can we set  `restock_refunded_items` to 0 for every new order whenever the first refund happens so that it's not undefined for new orders?

roykho replied:
So it won't be undefined because we're wrapping it in `wc_stock_amount()` which evaluate it to `0` if it is empty. I have pushed up the changes per our discussion. I have also updated the test to work with WC 5.4+ only. The only thing new here is when you test this PR, you have to set https://github.com/woocommerce/woocommerce/blob/trunk/includes/class-woocommerce.php#L27 to version `5.4`

vedanshujain commented on includes/wc-order-functions.php:
```diff
@@ -715,9 +715,13 @@ function wc_restock_refunded_items( $order, $refunded_line_items ) {
+			// Keeps track of total running tally of reduced stock.
 			$item->update_meta_data( '_reduced_stock', $item_stock_reduced );
+			// Keeps track of only refunded items that needs restock.
+			$item->update_meta_data( '_restock_refunded_items', $qty_to_refund );
```
Shouldn't this be an addition op instead of an update? So if we are tracking the refunded quantity we are supposed to stock back, we would add to the already existing quantity right, instead of updating it? I think in the current code there would be a bug when we :
1. Refund with restock
2. Refund with restock again
3. Update order

It will only keep track of the last restock operation, and will reduce the stock again from our first refund.

vedanshujain reviewed (changes requested):
Left some question, also since this is a pretty complicated part of codebase, it will be good to add unit tests here.

roykho committed:
Add to the restock_refunded_items count instead of replacting

roykho committed:
Add unit tests

roykho committed:
Update value of  to take refunded item quantity when  is not set

roykho committed:
Limit the new refund restock logic to order version 5.4+

roykho committed:
Restore testing version

roykho committed:
Use proper skip test method instead of return

roykho committed:
Initialize _restock_refunded_items meta on order update

roykho committed:
Fix merge conflicts

roykho committed:
Remove unneeded order_version

roykho commented:
Hi @vedanshujain - ok I have updated this PR based on our conversation. The change is that I've added a "on-the-fly" update when the order is saved in the backend. When that happens, it adds the `_restock_refunded_items` meta to orders that are created older versions (<5.4) and initialize it.

This should then retain the same behavior to older created orders while for newly created orders, it should count the restocked items correctly. Please have another look.

vedanshujain commented on src/Internal/RestockRefundedItemsAdjuster.php:
```diff
@@ -0,0 +1,67 @@
+			return;
+		}
+
+		if ( isset( $items['order_item_id'] ) ) {
```
perhaps we can bail early when there are no refunds to save processing time as well.

roykho replied:
Hmmm I was under the impression that we wanted to initialize this for all older orders that gets updated again? Otherwise we wouldn't need to do this here and could have just initialized it to zero if it doesn't exists which we saw it doesn't work as we know newer orders can have zero as legitimate value? Please elaborate more on this in case I am missing something.

vedanshujain replied:
sorry, i meant that in the loop below, we are iterating over every line item and doing an insert. so I was just wondering, that if there are no refunds at all on the order that we are saving (whether stocked or not), then we can just bail early and save on that processing and make the request faster.

roykho replied:
Perhaps you would like to make a suggestion commit?

vedanshujain reviewed (commented):
i have reviewed the code and leaving a minor comment now. i am still testing, but posting the review now so if everything works well, I can merge it and we will make it to code freeze.

vedanshujain commented on src/Internal/RestockRefundedItemsAdjuster.php:
```diff
@@ -0,0 +1,67 @@
+		if ( version_compare( $order_version, '5.4', '>=' ) ) {
+			return;
+		}
+
```
```suggestion
       // If there are no refund lines, then this migration isn't necessary because restock related meta's wouldn't be set.
		if ( 0 === count( $order->get_refunds() ) ) {
			return;
		}
```

```

vedanshujain commented on src/Internal/RestockRefundedItemsAdjuster.php:
```diff
@@ -0,0 +1,67 @@
+				if ( 'line_item' !== $item->get_type() ) {
+					continue;
+				}
+
```
```suggestion
    // there could be code paths in custom code which don't update version number but still update the items.
	if ( '' !== $item->get_meta( '_restock_refunded_items', true ) ) {
					continue;
	}
```

vedanshujain reviewed (commented):
Added code suggestions, along with one more question that I had. seems like if we delete order items, then we call `wc_maybe_adjust_line_item_product_stock` but not `woocommerce_before_save_order_items` so our hook may need to change.

roykho committed:
Update src/Internal/RestockRefundedItemsAdjuster.php

Co-authored-by: Vedanshu Jain <vedanshu.jain.2012@gmail.com>

roykho committed:
Update src/Internal/RestockRefundedItemsAdjuster.php

Co-authored-by: Vedanshu Jain <vedanshu.jain.2012@gmail.com>

roykho committed:
Clean up code and update version

roykho commented on src/Internal/RestockRefundedItemsAdjuster.php:
```diff
@@ -0,0 +1,67 @@
+	 */
+	final public function init() {
+		$this->order_factory = wc_get_container()->get( LegacyProxy::class )->get_instance_of( \WC_Order_Factory::class );
+		add_action( 'woocommerce_before_save_order_items', array( $this, 'initialize_restock_refunded_items' ), 10, 2 );
```
So I have pushed up the changes you suggested. I've tested every possible scenario I can think of and so far it is working the way we want. As for your comment about the hook, I think it would be fine. Even if they called `wc_maybe_adjust_line_item_product_stock` directly it wouldn't effect anything as `_restock_refunded_items` initializes to zero if it isn't set. Because I think the intent is set that meta as soon as we can for older order versions.

vedanshujain reviewed (approved):
LGTM! Great work!
